Mathlib/GroupTheory/GroupAction/Blocks.lean

```diff
@@ -470,6 +485,109 @@ def block_stabilizerOrderIso [htGX : IsPretransitive G X] (a : X) :
 
 end Stabilizer
 
+section Finite
+
+namespace IsBlock
+
+variable [IsPretransitive G X] {B : Set X} 
+
+theorem ncard_block_eq_relindex (hB : IsBlock G B) {x : X} (hx : x ∈ B) :
+    B.ncard = (stabilizer G x).relindex (stabilizer G B) := by
+  have key : (stabilizer G x).subgroupOf (stabilizer G B) = stabilizer (stabilizer G B) x := by
+    ext; rfl
+  rw [Subgroup.relindex, key, index_stabilizer, hB.orbit_stabilizer_eq hx]
+
+/-- The cardinality of the ambient is the product of
+  of the cardinality of a block
+  by the cardinality of the set of translates of that block -/
+theorem ncard_block_mul_ncard_orbit_eq (hB : IsBlock G B) (hB_ne : B.Nonempty) :
+    Set.ncard B * Set.ncard (orbit G B) = Nat.card X := by
+  obtain ⟨x, hx⟩ := hB_ne
+  rw [ncard_block_eq_relindex hB hx, ← index_stabilizer,
+      Subgroup.relindex_mul_index (hB.stabilizer_le hx), index_stabilizer']
+
+/-- The cardinality of a block divides the cardinality of the ambient type -/
+theorem ncard_dvd_card (hB : IsBlock G B) (hB_ne : B.Nonempty) :
+    Set.ncard B ∣ Nat.card X :=
+  Dvd.intro _ (hB.ncard_block_mul_ncard_orbit_eq hB_ne)
+
+/-- A too large block is equal to ⊤ -/
+theorem eq_univ_card_lt [hX : Finite X] (hB : IsBlock G B) (hB' : Nat.card X < Set.ncard B * 2) :
+    B = Set.univ := by
+  rcases Set.eq_empty_or_nonempty B with rfl | hB_ne
+  · simp only [Set.ncard_empty, zero_mul, not_lt_zero'] at hB'
+  have key := hB.ncard_block_mul_ncard_orbit_eq hB_ne
+  rw [← key, mul_lt_mul_iff_of_pos_left (by rwa [Set.ncard_pos])] at hB'
+  interval_cases (orbit G B).ncard
+  · rw [mul_zero, eq_comm, Nat.card_eq_zero, or_iff_left hX.not_infinite] at key
+    exact (IsEmpty.exists_iff.mp hB_ne).elim
+  · rw [mul_one, ← Set.ncard_univ] at key
+    rw [Set.eq_of_subset_of_ncard_le (Set.subset_univ B) key.ge]
+
+/-- If a block has too many translates, then it is a (sub)singleton  -/
+theorem subsingleton_of_card_lt [Finite X] (hB : IsBlock G B)
+    (hB' : Nat.card X < 2 * Set.ncard (Set.range fun g : G => (g • B : Set X))) :
+    B.Subsingleton := by
+  suffices Set.ncard B < 2 by
+    rw [Nat.lt_succ_iff, Set.ncard_le_one_iff_eq] at this
+    cases this with
+    | inl h => rw [h]; exact Set.subsingleton_empty
+    | inr h =>
+      obtain ⟨a, ha⟩ := h; rw [ha]; exact Set.subsingleton_singleton
+  cases Set.eq_empty_or_nonempty B with
+  | inl h => rw [h, Set.ncard_empty]; norm_num
+  | inr h =>
+    rw [← hB.ncard_block_mul_ncard_orbit_eq h, lt_iff_not_ge] at hB'
+    rw [← not_le]
+    exact fun hb ↦ hB' (Nat.mul_le_mul_right _ hb)
+
+-- TODO : Is the assumption B.finite necessary ?
```

Comment:
I think it is necessary. If you start the proof with the reversible steps
```
  rw [mk_subset]
  intro g x h1 h2 y h3
  obtain ⟨gx, rfl⟩ := exists_smul_eq G a x
  obtain ⟨gy, rfl⟩ := exists_smul_eq G a y
  simp only [Set.mem_smul_set_iff_inv_smul_mem, Set.mem_iInter] at h1 h2 h3 ⊢
```
then you get the goal state
```
B : Set X
a : X
g gx gy : G
h1 : ∀ (i : G), i⁻¹ • a ∈ B → i⁻¹ • gx • a ∈ B
h2 : ∀ (i : G), i⁻¹ • a ∈ B → i⁻¹ • g⁻¹ • gx • a ∈ B
h3 : ∀ (i : G), i⁻¹ • a ∈ B → i⁻¹ • g⁻¹ • gy • a ∈ B
⊢ ∀ (i : G), i⁻¹ • a ∈ B → i⁻¹ • gy • a ∈ B
```
which cannot be proven without further assumptions: `h1` lets you right-multiply by `gx`, `h2` lets you right-multiply by `g⁻¹ * gx`, `h3` lets you right-multiply by `g⁻¹ * gy`, but together these do not let you right-multiply by `gy`. You need some sort of finiteness or periodicity so that you can (for instance) right multiply by `gx * (g⁻¹ * gx)⁻¹ * (g⁻¹ * gy)`.

I think a counterexample might be `G = ℤ` acting on `ℤ` by translation, `B = ℕ`, and `a = 0`.